src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs

```diff
@@ -330,54 +330,98 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
         }
 
-        private static IEnumerable<ResolverPackage> FindCircularDependency(ResolverPackage package, IEnumerable<string> parents, IEnumerable<ResolverPackage> solution)
+        private static List<ResolverPackage> FindCircularDependency(ResolverPackage package, Dictionary<string, ResolverPackage> packageLookUp, HashSet<ResolverPackage> visitedPackages)
         {
             // avoid checking depths beyond 20 packages deep
-            if (parents.Count() < 20 && package != null && !package.Absent && package.Dependencies.Any())
-            {
-                // walk the dependencies
-                foreach (var dependency in package.Dependencies.OrderBy(d => d.Id, StringComparer.OrdinalIgnoreCase))
+            //ashish: removed 20 packages check to go by any depth since no more stack processing
+            if (package != null && !package.Absent && package.Dependencies.Any())
+            {                
+                var queue = new Queue<QueueNode>();
+                //added the first initial node to queue
+                var qNode = new QueueNode(package, new List<ResolverPackage>());
+                queue.Enqueue(qNode);
+
+                //BFS traversal to traverse through all the packages to find out circular dependency
+                while (queue.Count > 0)
                 {
-                    var dependencyPackage = solution.FirstOrDefault(solutionPackage => StringComparer.OrdinalIgnoreCase.Equals(solutionPackage.Id, dependency.Id));
+                    var source = queue.Dequeue();
+                    //access parent list and add current package as well
+                    var parentPackages = new List<ResolverPackage>(source.Parent);
+                    parentPackages.Add(source.Package);
 
-                    if (parents.Contains(dependency.Id, StringComparer.OrdinalIgnoreCase))
+                    // walk the dependencies
+                    foreach (var dependency in source.Package.Dependencies.OrderBy(d => d.Id, StringComparer.OrdinalIgnoreCase))
                     {
-                        // loop detected
-                        return new ResolverPackage[] { package, dependencyPackage };
-                    }
-
-                    // recurse on dependencies
-                    var result = FindCircularDependency(dependencyPackage, parents.Concat(new string[] { package.Id }), solution);
-
-                    if (result.Any())
-                    {
-                        return (new ResolverPackage[] { package }).Concat(result);
+                        var dependencyPackage = packageLookUp[dependency.Id];
+                        //If already visited, then check current source package for circular dependency since it means we've already processed child nodes
+                        //bottom-up approach                     
+                        if (!visitedPackages.Contains(dependencyPackage))
+                        {
+                            //top-down approach
```

Comment:
I don't understand the bottom=up and top-down comments, either clarify or remove